[extension/oauth2clientauth] Enable dynamically reading ClientID and ClientSecret from files

elikatsis opened the pull request:
**Description:**
This PR implements the feature described in detail in the issue linked below.

In a nutshell, it extends the `oauth2clientauth` extension to read ClientID and/or ClientSecret from files whenever a new token is needed for the OAuth flow.
As a result, the extension can use updated credentials (when the old ones expire for example) without the need to restart the OTEL collector, as long as the file contents are in sync.

**Link to tracking Issue:** #26117 

**Testing:**
Apart from the unit testing you can see in the PR, I've tested this feature in two real-life environments:

1. As a systemd service exporting `otlphttp` data
2. A Kubernetes microservice (deployed by an OpenTelemetryCollector CR) exporting `otlphttp` data

In both cases, the collectors export the data to a service which sits behind an OIDC authentication proxy.
Using the `oauth2clientauth` extension, the `otlphttp` exporter hits the authentication provider to issue tokens for the OIDC client and successfully authenticates to the service.

In my cases, the ClientSecret gets rotated quite frequently and there is a stack making sure the ClientID and ClientSecret in the corresponding files are up-to-date.

**Documentation:**
I have extended the extension's README file. I'm open to more suggestions!

cc @jpkrohling @pavankrish123

jpkrohling commented:
Can you sort out the CLA before I review this one?

elikatsis commented:
@jpkrohling CLA :white_check_mark:

elikatsis committed:
Implement wrapper for clientcredentials.Config

Implement a wrapper for clientcredentials.Config in preparation for
custom behavior and logic when issuing new tokens.

elikatsis committed:
Enable specifying files to get client ID/secret from

Leverage the clientCredentialsConfig features to read client ID and/or
secret from a file to the actual extension configuration.

elikatsis committed:
Add Changelog entry file

elikatsis committed:
Extend creds config to retrieve client ID/secret from files

Extend the clientCredentailsConfig to retrieve client ID and/or secret
from files by setting values to the ClientIDFile and ClientSecretFile
fields respectively.

The Client*File field takes precedence over the Client* field.

elikatsis committed:
Add tests for ClientIDFile and ClientSecretFile fields

elikatsis committed:
Use local clientCredentialsConfig

Use the locally defined struct clientCredentialsConfig in favor of the
upstream clientcredentials.Config object.

jpkrohling commented on extension/oauth2clientauthextension/clientcredentialsconfig.go:
```diff
@@ -0,0 +1,105 @@
+	var err error
+
+	if len(filepath) > 0 {
+		actualValue, err = readCredentialsFile(filepath)
```
Do you need to create a new var? Perhaps just return the results of the readCredentialsFile if the filepath's len is > 0, otherwise return value, nil ?

```suggestion
		return readCredentialsFile(filepath)
```

elikatsis replied:
At this point we don't create a new var, we override the `actualValue` one (that's why we also define `err` outside the block). This doesn't look optimal at this state (and I'm good with changing it) but it makes things more clear in my opinion when adding more ways of retrieving the "actual value" (as mentioned in the issue, get by executing a command).

I can push the commits with the command part so that you can take a look and decide how you want me to change it.

elikatsis replied:
@jpkrohling this is the relevant commit adding the command case: https://github.com/elikatsis/opentelemetry-collector-contrib/commit/e453109b958265f58ca71b37701a5b7797469b18

jpkrohling replied:
> as mentioned in the issue, get by executing a command

I missed that, and this specific example is unlikely to get accepted: we probably do not want the collector executing commands.

> At this point we don't create a new var, we override the actualValue one

I mean something like this:

```golang
func getActualValue(value, cmd, filepath string) (string, error) {
	if len(filepath) > 0 {
		return readCredentialsFile(filepath)
	}

	if len(cmd) > 0 {
		return readCommandOutput(cmd)
	}

	return value, nil
}
```

jpkrohling replied:
By the way, this is a minor improvement, optional for the approval of this PR.

Another take would be this: given that this is in the hot path and the decision will be the same during the whole lifecycle of this component, it could even be better to just use a func as a parameter to this type. Like this:

```golang
func NewClientCredentialsConfig(cfg *clientcredentials.Config, clientIDFile, clientSecretFile string) (*clientCredentialsConfig, error) {
	c := &clientCredentialsConfig{
		Config: *cfg,
	}

	if len(clientIDFile) > 0 {
		c.clientIDSource = func() (string, error) {
			return readCredentialsFile(clientIDFile)
		}
	} else {
		c.clientIDSource = func() (string, error) {
			return c.ClientID, nil
		}
	}

	if len(clientSecretFile) > 0 {
		c.clientSecretSource = func() (string, error) {
			return readCredentialsFile(clientSecretFile)
		}
	} else {
		c.clientSecretSource = func() (string, error) {
			return c.ClientSecret, nil
		}
	}

	return c, nil
}
```

Your createConfig would then call `clientID, err := c.clientIDSource()` and `clientSecret, err := c.clientSecretSource()` accordingly.

jpkrohling commented on extension/oauth2clientauthextension/clientcredentialsconfig.go:
```diff
@@ -0,0 +1,105 @@
+//		},
+//		ClientSecretFile: "/path/to/client/secret",
+//	}
+type clientCredentialsConfig struct {
```
Can't clientcredentials.Config be made to implement TokenSource instead of having a separate struct for this? I feel like this is leaking some concerns the caller shouldn't have.

elikatsis replied:
I'm not sure I get what you're saying. The API is for the user (here, the extension) to define a "Config". Here's my take:

* We need a new Config object to include extra options (`ClientIDFile`, `ClientSecretFile`)
* We need a TokenSource object on top of our config to essentially produce a proper Token by coming up with a proper `clientcredentials.Config`

Does this cover your question? Otherwise, could you elaborate?

elikatsis replied:
As mentioned in the issue the API derived from the upstream Go module is
* Have a Config object with all the configurations needed
* Produce a TokenSource object from the Config (using a Config method)
* Generate Tokens from the TokenSource object (using a TokenSource method)

jpkrohling commented on extension/oauth2clientauthextension/extension.go:
```diff
@@ -55,12 +55,16 @@ func newClientAuthenticator(cfg *Config, logger *zap.Logger) (*clientAuthenticat
-			Scopes:         cfg.Scopes,
-			EndpointParams: cfg.EndpointParams,
+		clientCredentials: &clientCredentialsConfig{
+			Config: clientcredentials.Config{
```
Here, it's more evident that the caller has to know more than it should about the config: why would it need to use a config within a config?

jpkrohling commented on extension/oauth2clientauthextension/extension_test.go:
```diff
@@ -134,6 +135,98 @@ func TestOAuthClientSettings(t *testing.T) {
+			cfg, err := rc.clientCredentials.createConfig()
+			if test.shouldError {
+				assert.NotNil(t, err)
+				assert.Contains(t, err.Error(), test.expectedError)
```
If you use `ErrorAs` here, you can use the actual error (`errNoClientSecretProvided`) instead of doing a string comparison with `errNoClientSecretProvided.Error()`.

elikatsis replied:
I followed the same pattern as other tests so that it's less things to review. Do you want me to change it as you suggest for the changes of my PR?

jpkrohling replied:
It's fine, it's just a minor comment.

jpkrohling commented on extension/oauth2clientauthextension/README.md:
```diff
@@ -74,7 +74,11 @@ Following are the configuration fields
 - [**token_url**](https://datatracker.ietf.org/doc/html/rfc6749#section-3.2) - The resource server's token endpoint URLs.
 - [**client_id**](https://datatracker.ietf.org/doc/html/rfc6749#section-2.2) - The client identifier issued to the client.
+- **client_id_file** - The file path to retrieve the client identifier issued to the client.
+  This setting takes precedence over `client_id`.
```
It would be nice to clarify how (and whether) the values are refreshed when the backing file has changed.

elikatsis replied:
Ack

jpkrohling reviewed (commented):
I have made a few suggestions but this looks very cool! Thank you for the PR.

elikatsis commented:
@jpkrohling thank you very much for your review. I've posted some replies I'd like your feedback on

elikatsis committed:
Address review comments

elikatsis commented:
@jpkrohling I made the following changes
* Extended README
* Modified tests to not use string comparisons to assert errors
* Changed `getActualValue` to immediately return instead of using extra vars

Your suggestion on using `client*Source()` methods is very interesting. However, I didn't make the change because it feels like it won't help much in case we need something similar for other attributes as well or introduce more options for these attributes. Having the logic to collect values in `createConfig()` makes it more easy and straight forward if anyone wants to extend it. I also think there isn't any actual performance improvement.
**If you prefer it, I will definitely do it,** no strong opinion :smiley:

jpkrohling reviewed (approved):
LGTM for a first implementation. The next iteration could perhaps use the same approach we have in the Jaeger Remote sampling extension, by watching the file for changes instead of re-reading on every request. I think there's a big performance penalty with this, but I'll accept because:

1) it's ready and tested by the author
2) it's a new feature, doesn't impact current users

elikatsis commented:
@jpkrohling I fixed a typo which made the lint test fail. I ran quite a few more Makefile targets trying to catch things early this time. You'll probably need to trigger the tests again :/

elikatsis commented:
I see this error in the test
```
[2023-09-06T17:00:51.035Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
```

Is it relevant to my PR or it's some infra issue? Hmu if I can do anything to unblock this

jpkrohling commented:
I ran it again and it's passing.
